File: src/main/java/org/parthvnp/Array/ThreeSumSmaller.java
package org.parthvnp.Array;

import java.util.*;


public class ThreeSumSmaller {

    public int threeSumSmaller(int[] nums, int target) {
        int count = 0;
        Arrays.sort(nums);
        for (int i = 0; i < nums.length - 2; i++) {
            int low = i + 1, high = nums.length - 1;
            while (low < high) {
                if (nums[i] + nums[low] + nums[high] < target) {
                    count += high - low;
                    low++;
                } else {
                    high--;
                }
            }
        }
        return count;
    }

    public static void main(String[] args) {
        var s = new ThreeSumSmaller();
        var e2 = s.threeSumSmaller(new int[]{3, 1, 0, -2}, 4);
        System.out.println(e2 == 3);
        System.out.println(e2);

    }
}

File: src/main/python/arrays/three_sum_smaller.py
from typing import *
from collections import *
from shared.test import test


class ThreeSumSmaller:
    def threeSumSmaller(self, nums: List[int], target: int) -> int:
        count = 0
        n = len(nums)
        nums.sort()
        for i in range(n - 2):
            low = i + 1
            high = n - 1
            while low < high:
                if nums[i] + nums[low] + nums[high] < target:
                    count += high - low
                    low += 1
                else:
                    high -= 1
        return count


if __name__ == "__main__":
    s = ThreeSumSmaller()
    e1 = s.threeSumSmaller([3, 4, 2], 6)
    test(0, e1)
    e2 = s.threeSumSmaller([3, 1, 0, -2], 4)
    test(3, e2)
    e3 = s.threeSumSmaller([1, 1, -2], 1)
    test(1, e3)
    e4 = s.threeSumSmaller([1, 1, -2], 0)
    test(0, e4)
